x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportClientEnrollmentAction.java

```diff
@@ -0,0 +1,128 @@
+
+public class TransportClientEnrollmentAction extends HandledTransportAction<ClientEnrollmentRequest, ClientEnrollmentResponse> {
+
+    private static final Logger logger = LogManager.getLogger(TransportClientEnrollmentAction.class);
+
+    private final Environment environment;
+    private final Client client;
+    private final Settings settings;
+    private final SSLService sslService;
+
+    @Inject
+    public TransportClientEnrollmentAction(TransportService transportService, Client client, Settings settings,
+                                                   SSLService sslService, Environment environment, ActionFilters actionFilters) {
+        super(ClientEnrollmentAction.NAME, transportService, actionFilters, ClientEnrollmentRequest::new);
+        this.environment = environment;
+        this.client = new OriginSettingClient(client, SECURITY_ORIGIN);
+        this.settings = settings;
+        this.sslService = sslService;
+    }
+
+    @Override
+    protected void doExecute(
+        Task task, ClientEnrollmentRequest request, ActionListener<ClientEnrollmentResponse> listener) {
+        try {
+            final KeyConfig keyConfig = sslService.getHttpTransportSSLConfiguration().keyConfig();
+            if (keyConfig instanceof StoreKeyConfig == false) {
+                listener.onFailure(new IllegalStateException(
+                        "Unable to enroll client. Elasticsearch node HTTP layer SSL configuration is not configured with a keystore"));
+                return;
+            }
+            final List<X509Certificate> caCertificates = ((StoreKeyConfig) keyConfig).x509Certificates(environment).stream()
+                    .filter(x509Certificate -> x509Certificate.getBasicConstraints() != -1).collect(Collectors.toList());
+            if (caCertificates.size() != 1) {
+                    listener.onFailure(new IllegalStateException(
+                        "Unable to enroll client. Elasticsearch node HTTP layer SSL configuration Keystore doesn't contain a single " +
+                            "PrivateKey entry where the associated certificate is a CA certificate"));
+
+            } else {
+                final String httpCa = Base64.getUrlEncoder().encodeToString(caCertificates.get(0).getEncoded());
+                final NodesInfoRequest nodesInfoRequest = new NodesInfoRequest().addMetric(NodesInfoRequest.Metric.HTTP.metricName());
+                client.execute(NodesInfoAction.INSTANCE, nodesInfoRequest, ActionListener.wrap(nodesInfoResponse -> {
+                    final List<String> nodeList = new ArrayList<>();
+                    for (NodeInfo nodeInfo : nodesInfoResponse.getNodes()) {
+                        nodeList.add(nodeInfo.getInfo(HttpInfo.class).getAddress().publishAddress().toString());
+                    }
+                    if (shouldSetPassword(request)) {
+                        final String user = userPasswordToSet(request);
+                        assert user != null;
```

Comment:
Should we throw an exception here?

Reply:
I added an assert here as this should _not_ happen, but wanted to check eitherway and for future proofing. `shouldSetPassword` should only return true when the client type is kibana and when the client type is kibana the `userPasswordToSet` is hardcoded to return `kibana_system`.